packages/core/js-client/src/compilerSupport/conversions.ts

```diff
@@ -13,59 +13,78 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { jsonify } from '../util/utils.js';
-import { match } from 'ts-pattern';
-import type { ArrowType, ArrowWithoutCallbacks, NonArrowType } from '@fluencelabs/interfaces';
-import { CallServiceData } from '../jsServiceHost/interfaces.js';
+
+// TODO: This file is a mess. Need to refactor it later
+/* eslint-disable */
+// @ts-nocheck
+
+import assert from "assert";
+
+import type {
+  ArrowType,
+  ArrowWithoutCallbacks,
+  JSONArray,
+  JSONValue,
+  NonArrowType,
+} from "@fluencelabs/interfaces";
+import { match } from "ts-pattern";
+
+import { CallServiceData } from "../jsServiceHost/interfaces.js";
+import { jsonify } from "../util/utils.js";
 
 /**
  * Convert value from its representation in aqua language to representation in typescript
  * @param value - value as represented in aqua
  * @param type - definition of the aqua type
  * @returns value represented in typescript
  */
-export const aqua2ts = (value: any, type: NonArrowType): any => {
-    const res = match(type)
-        .with({ tag: 'nil' }, () => {
-            return null;
-        })
-        .with({ tag: 'option' }, (opt) => {
-            if (value.length === 0) {
-                return null;
-            } else {
-                return aqua2ts(value[0], opt.type);
-            }
-        })
-        // @ts-ignore
-        .with({ tag: 'scalar' }, { tag: 'bottomType' }, { tag: 'topType' }, () => {
-            return value;
-        })
-        .with({ tag: 'array' }, (arr) => {
-            return value.map((y: any) => aqua2ts(y, arr.type));
-        })
-        .with({ tag: 'struct' }, (x) => {
-            return Object.entries(x.fields).reduce((agg, [key, type]) => {
-                const val = aqua2ts(value[key], type);
-                return { ...agg, [key]: val };
-            }, {});
-        })
-        .with({ tag: 'labeledProduct' }, (x) => {
-            return Object.entries(x.fields).reduce((agg, [key, type]) => {
-                const val = aqua2ts(value[key], type);
-                return { ...agg, [key]: val };
-            }, {});
-        })
-        .with({ tag: 'unlabeledProduct' }, (x) => {
-            return x.items.map((type, index) => {
-                return aqua2ts(value[index], type);
-            });
-        })
-        // uncomment to check that every pattern in matched
-        // .exhaustive();
-        .otherwise(() => {
-            throw new Error('Unexpected tag: ' + jsonify(type));
-        });
-    return res;
+export const aqua2ts = (value: JSONValue, type: NonArrowType): JSONValue => {
+  const res = match(type)
+    .with({ tag: "nil" }, () => {
+      return null;
+    })
+    .with({ tag: "option" }, (opt) => {
+      assert(Array.isArray(value));
```

Comment:
if you wanna do it in separate PR that's ok. I just would remove these new asserts for now becuase it would be good to have a propper error message if this assert is possible and if it should be impossimble - good error message that explains why this error should never happen would be uselful as well